src/server/pfs/server/driver.go

```diff
@@ -675,6 +671,66 @@ func (d *driver) deleteProject(ctx context.Context, txnCtx *txncontext.Transacti
 	return nil
 }
 
+// Set child.ParentCommit (if 'parent' has been determined) and write child to parent's ChildCommits
+func (d *driver) linkParent(txnCtx *txncontext.TransactionContext, child *pfs.CommitInfo, parent *pfs.Commit, needsFinishedParent bool) error {
+	if parent == nil {
+		return nil
+	}
+	// Resolve 'parent' if it's a branch that isn't 'branch' (which can
+	// happen if 'branch' is new and diverges from the existing branch in
+	// 'parent').
+	// Clone the parent proto because resolveCommit will modify it.
+	parent = proto.Clone(parent).(*pfs.Commit)
+	parentCommitInfo, err := d.resolveCommit(txnCtx.SqlTx, parent)
+	if err != nil {
+		return errors.Wrapf(err, "parent commit not found")
+	}
+	// fail if the parent commit has not been finished
+	if needsFinishedParent && parentCommitInfo.Finishing == nil {
+		return errors.Errorf("parent commit %s has not been finished", parentCommitInfo.Commit)
+	}
+	child.ParentCommit = parentCommitInfo.Commit
+	parentCommitInfo.ChildCommits = append(parentCommitInfo.ChildCommits, child.Commit)
+	// Note: error is emitted if parent.ID is a missing/invalid branch OR a
+	// missing/invalid commit ID
+	return errors.Wrapf(d.commits.ReadWrite(txnCtx.SqlTx).Put(parentCommitInfo.Commit, parentCommitInfo),
+		"could not resolve parent commit %s", parent)
+}
+
+// TODO(acohen4): reassess need for "needsFinishedParent" parameter
+// creates a new commit, and adds both commit ancestry, and commit provenance pointers
+func (d *driver) addCommit(txnCtx *txncontext.TransactionContext, newCommitInfo *pfs.CommitInfo, parent *pfs.Commit, directProvenance []*pfs.Branch, needsFinishedParent bool) error {
+	if err := d.linkParent(txnCtx, newCommitInfo, parent, needsFinishedParent); err != nil {
+		return err
+	}
+	provHeads := make(map[string]*pfs.Commit)
+	for _, prov := range directProvenance {
+		branchInfo := &pfs.BranchInfo{}
+		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(pfsdb.BranchKey(prov), branchInfo); err != nil {
+			if col.IsErrNotFound(err) {
+				return pfsserver.ErrBranchNotFound{Branch: prov}
+			}
+			return errors.EnsureStack(err)
+		}
+		provHeads[pfsdb.CommitKey(branchInfo.Head)] = branchInfo.Head
+	}
+	for _, p := range provHeads {
+		newCommitInfo.DirectProvenance = append(newCommitInfo.DirectProvenance, p)
```

Comment:
We should keep the direct provenance order stable when going from branches to commits. I don't think the deduplication is needed anymore since multiple branches in the same repo cannot be in the same provenance DAG.